File: app/src/main/java/com/lumiyaviewer/lumiya/render/glres/buffers/GLVertexArrayObject.java
package com.lumiyaviewer.lumiya.render.glres.buffers;

import android.annotation.TargetApi;
import android.opengl.GLES30;
import com.lumiyaviewer.lumiya.render.glres.GLGenericResource;
import com.lumiyaviewer.lumiya.render.glres.GLResourceManager;

@TargetApi(18)
public class GLVertexArrayObject implements GLGenericResource {
    public final int size;
    private final int[] vaoIndices;

    private static class GLVertexArrayObjectReference extends GLResourceManager.GLGenericResourceReference {
        private final int[] vaoIndices;

        GLVertexArrayObjectReference(GLGenericResource glGenericResource, GLResourceManager glResourceManager, int[] ints) {
            super(glGenericResource, glResourceManager);
            this.vaoIndices = ints;
        }

        @Override
        public void GLFree() {
            GLES30.glDeleteVertexArrays(this.vaoIndices.length, this.vaoIndices, 0);
        }
    }

    @TargetApi(18)
    public GLVertexArrayObject(GLResourceManager glResourceManager, int size) {
        this.size = size;
        this.vaoIndices = new int[size];
        GLES30.glGenVertexArrays(size, this.vaoIndices, 0);
        new GLVertexArrayObjectReference(this, glResourceManager, this.vaoIndices);
    }

    public void Bind(int i) {
        GLES30.glBindVertexArray(this.vaoIndices[i]);
    }

    public void Unbind() {
        GLES30.glBindVertexArray(0);
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/render/glres/buffers/GLVertexArrayObject.kt
package com.lumiyaviewer.lumiya.render.glres.buffers

import android.opengl.GLES30
import com.lumiyaviewer.lumiya.render.glres.GLGenericResource
import com.lumiyaviewer.lumiya.render.glres.GLResourceManager

class GLVertexArrayObject(glResourceManager: GLResourceManager, val size: Int) : GLGenericResource {
    private val vaoIndices: IntArray = IntArray(size)

    private class GLVertexArrayObjectReference(
        glGenericResource: GLGenericResource,
        glResourceManager: GLResourceManager,
        private val vaoIndices: IntArray
    ) : GLResourceManager.GLGenericResourceReference(glGenericResource, glResourceManager) {
        override fun GLFree() {
            GLES30.glDeleteVertexArrays(vaoIndices.size, vaoIndices, 0)
        }
    }

    init {
        GLES30.glGenVertexArrays(size, vaoIndices, 0)
        GLVertexArrayObjectReference(this, glResourceManager, vaoIndices)
    }

    fun Bind(i: Int) {
        GLES30.glBindVertexArray(vaoIndices[i])
    }

    fun Unbind() {
        GLES30.glBindVertexArray(0)
    }
}
